File: Baekjoon/16562 - 친구비/16562.cpp
#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;

#define FASTIO cin.tie(NULL); cout.tie(NULL); ios::sync_with_stdio(false);
#define INF 1000000000

vector<int> parent(1, 0);

int Find(int x) {
    if(x == parent[x]) return x;
    else return parent[x] = Find(parent[x]);
}

void Union(int x, int y) {
    x = Find(x);
    y = Find(y);

    if(x > y) parent[x] = y;
    else parent[y] = x;
}

int main() {
    FASTIO
    
    int n, m, k, a, b, temp, result = 0;
    vector<int> cost(1, 0), last(1, INF);

    cin >> n >> m >> k;

    for(int i = 1; i <= n; i++) {
        cin >> temp;

        cost.push_back(temp);
        last.push_back(INF);
        parent.push_back(i);
    }

    for(int _ = 0; _ < m; _++) {
        cin >> a >> b;
        Union(a, b);
    }

    for(int i = 1; i <= n; i++) {
        parent[i] = Find(i);
        
        if(last[parent[i]] == INF) {
            result += cost[i];
            last[parent[i]] = cost[i];
        }
        else if(last[parent[i]] != INF && last[parent[i]] > cost[i]) {
            result -= last[parent[i]] - cost[i];
            last[parent[i]] = cost[i];
        }
    }

    if(result <= k) cout << result;
    else cout << "Oh no";
}

File: solution/16xxx/16562 - 친구비/16562.py
import sys
input = sys.stdin.readline

def find(x, parent):
    if x != parent[x]:
        parent[x] = find(parent[x], parent)
    
    return parent[x]

def union(x, y, parent):
    x = find(x, parent)
    y = find(y, parent)

    if x > y:
        parent[x] = y
    else:
        parent[y] = x

n, m, k = map(int, input().split())
A = [0] + list(map(int, input().split()))

parent = []
cost = []

for i in range(0, n+1):
    parent.append(i)
    cost.append(1e9)

for _ in range(m):
    a, b = map(int, input().split())
    union(a, b, parent)

result = 0

for i in range(1, n+1):
    parent[i] = find(i, parent)

    if cost[parent[i]] == 1e9:
        result += A[i]
        cost[parent[i]] = A[i]
    elif cost[parent[i]] != 1e9 and cost[parent[i]] > A[i]:
        result -= cost[parent[i]] - A[i]
        cost[parent[i]] = A[i]

print(result) if result <= k else print("Oh no")